### ISSUE TITLE
CoreMIDI4J doesn't like repeatedly disconnecting devices

### ISSUE THREAD

**Issue by eclab** [DerekCook/CoreMidi4J#19]
I'm getting a lot of crashes after disconnecting devices and reconnecting again (perhaps 3 or 4 times -- not many!).  This is in Edisyn (https://github.com/eclab/edisyn/), from selecting the "MIDI:Change MIDI" menu 3 or 4 times.  Try it out!  I'm running Java 1.7.0_51 on OS X 10.9.5. I'll usually get a hang or a hard VM dump. Example VM dump stack trace:

```
uk.co.xfactorylibrarians.coremidi4j.CoreMidiException: Exception in CoreMIDI JNI Library by "MIDIPortDisconnectSource" - OS STatus Code: ffffd5af
	at uk.co.xfactorylibrarians.coremidi4j.CoreMidiInputPort.midiPortDisconnectSource(Native Method)
	at uk.co.xfactorylibrarians.coremidi4j.CoreMidiInputPort.disconnectSource(CoreMidiInputPort.java:71)
	at uk.co.xfactorylibrarians.coremidi4j.CoreMidiSource.close(CoreMidiSource.java:130)
	at edisyn.Synth.gatherMidiDevices(Synth.java:1203)
	at edisyn.Synth.setupMIDI(Synth.java:486)
	at edisyn.Synth.setupMIDI(Synth.java:468)
	at edisyn.Synth$14.actionPerformed(Synth.java:780)
	at javax.swing.AbstractButton.fireActionPerformed(AbstractButton.java:2018)
	at javax.swing.AbstractButton$Handler.actionPerformed(AbstractButton.java:2341)
	at javax.swing.DefaultButtonModel.fireActionPerformed(DefaultButtonModel.java:402)
	at javax.swing.DefaultButtonModel.setPressed(DefaultButtonModel.java:259)
	at javax.swing.AbstractButton.doClick(AbstractButton.java:376)
	at com.apple.laf.ScreenMenuItem.actionPerformed(ScreenMenuItem.java:128)
	at java.awt.MenuItem.processActionEvent(MenuItem.java:669)
	at java.awt.MenuItem.processEvent(MenuItem.java:628)
	at java.awt.MenuComponent.dispatchEventImpl(MenuComponent.java:351)
	at java.awt.MenuComponent.dispatchEvent(MenuComponent.java:339)
	at java.awt.EventQueue.dispatchEventImpl(EventQueue.java:738)
	at java.awt.EventQueue.access$200(EventQueue.java:103)
	at java.awt.EventQueue$3.run(EventQueue.java:694)
	at java.awt.EventQueue$3.run(EventQueue.java:692)
	at java.security.AccessController.doPrivileged(Native Method)
	at java.security.ProtectionDomain$1.doIntersectionPrivilege(ProtectionDomain.java:76)
	at java.security.ProtectionDomain$1.doIntersectionPrivilege(ProtectionDomain.java:87)
	at java.awt.EventQueue$4.run(EventQueue.java:708)
	at java.awt.EventQueue$4.run(EventQueue.java:706)
	at java.security.AccessController.doPrivileged(Native Method)
	at java.security.ProtectionDomain$1.doIntersectionPrivilege(ProtectionDomain.java:76)
	at java.awt.EventQueue.dispatchEvent(EventQueue.java:705)
	at java.awt.EventDispatchThread.pumpOneEventForFilters(EventDispatchThread.java:242)
	at java.awt.EventDispatchThread.pumpEventsForFilter(EventDispatchThread.java:161)
	at java.awt.EventDispatchThread.pumpEventsForHierarchy(EventDispatchThread.java:150)
	at java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:146)
	at java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:138)
	at java.awt.EventDispatchThread.run(EventDispatchThread.java:91)
#
# A fatal error has been detected by the Java Runtime Environment:
#
#  SIGBUS (0xa) at pc=0x00007fff754e8340, pid=18392, tid=77059
#
# JRE version: Java(TM) SE Runtime Environment (7.0_51-b13) (build 1.7.0_51-b13)
# Java VM: Java HotSpot(TM) 64-Bit Server VM (24.51-b03 mixed mode bsd-amd64 compressed oops)
# Problematic frame:
# C  [CoreFoundation+0xe6899340]  OBJC_METACLASS_$___NSCFString+0x0
#
# Failed to write core dump. Core dumps have been disabled. To enable core dumping, try "ulimit -c unlimited" before starting Java again
#
# An error report file with more information is saved as:
# /Users/sean/java/distros/edisyn.git/trunk/hs_err_pid18392.log
#
# If you would like to submit a bug report, please visit:
#   http://bugreport.sun.com/bugreport/crash.jsp
#
```

**Comment by eclab** [comment]
Forgot to include the err log.
[err.log.zip](https://github.com/DerekCook/CoreMidi4J/files/1036940/err.log.zip)

**Comment by eclab** [comment]
An additional bug, but one that I cannot produce a log for.  It looks like you can't dynamically allocate receivers.

1. I allocate three devices via uk.co.xfactorylibrarians.coremidi4j.CoreMidiDeviceProvider.getMidiDeviceInfo();

2. I open a transmitter on one device, and a receiver on two others.  

3. I use them a bit, then close the receivers and transmitters.  

4. I then open a new transmitter on one device, and a receiver on two others.  

5. On opening one of the receivers, the entire system reliably hangs hard and permanently.

Due to this and the previous bug, at present I have no choice but to create devices initially, then create hand-coded pipes for the receivers and transmitters that manage opening and closing.  This works but means that I cannot dynamically handle new devices coming online if the user plugs a synth into his computer after the software has been launched.

**Comment by DerekCook** [comment]
Hi, Sorry for the delay in replying. I am away on business and working manic hours, so not sure when I will get to look at this. It would be helpful if you can provide some sample code that guarantees both of the crashes you are talking about to give us something to work with in addition to the descriptions that you have provided.

**Comment by brunchboy** [comment]
Yes, sorry, I am incredibly buried with my other projects as well, and don’t know when I might have time to look at issues like this, but sample code that reproduces it reliably will definitely be key. Derek, if you do happen to find and fix it, I will of course make time to publish a new release.

**Comment by BradleyRoss** [comment]
Just a thought.  When you close all the transmitters and receivers, make sure that the devices are open before opening the new transmitters and receivers.  (MidiDevice.isOpen())  I saw something that indicated that the devices may close if the receivers and transmitters are closed.  I'm curious and will try running your project.

**Comment by eclab** [comment]
I'll try to get you guys a simple piece of example code [I've since modified my program heavily to work around the issue, so it won't be of much use to you except in early versions on github].

Sean

**Comment by eclab** [comment]
[Example.zip](https://github.com/DerekCook/CoreMidi4J/files/1086069/Example.zip)

On Jun 6, 2017, at 7:22 AM, James Elliott <notifications@github.com> wrote:

> Yes, sorry, I am incredibly buried with my other projects as well, and don’t know when I might have time to look at issues like this, but sample code that reproduces it reliably will definitely be key. Derek, if you do happen to find and fix it, I will of course make time to publish a new release.

I am attaching an example stripped out of older code I had.  It's not minimal but should be enough to demonstrate what's going on.  The code gathers all the available MIDI devices for a receiver and two transmitters, then attaches receivers to the transmitters.  It does this three times.  It then closes everything cleanly, at which point it crashes hard.  The code *should* be closing the existing receivers/transmitters and devices when it opens new ones.  I could try stripping it down further if need be.

**Comment by BradleyRoss** [comment]
To eclab

I don't see the attached code

**Comment by eclab** [comment]
Sorry, it looks like github silently removes files with "unsupported extensions" from email messages.  :-(  And for some reason I cannot fathom, .java is an unsupported extension.  Try now.

**Comment by BradleyRoss** [comment]
For one thing, you are testing if the various objects are not null, but I think you also need to check if the devices, transmitters, and receivers are open.  The synchronized lock doesn't appear right.  See http://winterbe.com/posts/2015/04/30/java8-concurrency-tutorial-synchronized-locks-examples/ and the java.util.concurrent.locks package.  It looks like it should be synchronized(this) according to the winterbe.com blog post.

I'm going to have to look at this in more detail, but that is the result of my first look.

See https://stackoverflow.com/questions/10963205/how-to-attach-file-to-a-github-issue  It appears that the acceptable file types are PNG, GIF, JPG, DOCX, PPTX, XLSX, TXT, or PDF

**Comment by eclab** [comment]
On Jun 20, 2017, at 3:08 AM, Bradley Ross <notifications@github.com> wrote:

> For one thing, you are testing if the various objects are not null, but I think you also need to check if the devices, transmitters, and receivers are open. 

They are null because I set them to be null immediately after closing them to make it clear that they're closed.  I think the code's right.


> The synchronized lock doesn't appear right.

You can comment out all the synchronized statements in the code because it's only single-threaded (its a remnant of my main code) but it won't make any difference.

I'm pretty sure the synchronization code is correct.  Locking on a lock object other than [this] is good practice for many purposes, particularly if you need to have independent locks for different methods.  Among the best choices for lock objects is an empty array because it is serializable (many other objects, including new Object(), are not).

**Comment by DerekCook** [comment]
Hi, Apologies for lack of response. I am am away on business and currently rammed. Any free time I have I have on programming is currently spent resolving an issue I have with one of my Java Librarians that use CoreMIDI4J. Will look at this as soon as I can, but if you can find the issue and generate a pull request then I am sure that James and I will be able to respond to that much more quickly. I am happy to consider extending who can actively contribute as well if you have the interest. CoreMIDI4J was put here so it could be supported unlike the older providers..

**Comment by brunchboy** [comment]
I should have some time to play with this over the weekend too, I hope. My massive updates to dysentery, beat-link, and beat-link-trigger are out in preview form, and feedback is settling down into a happy place.

**Comment by brunchboy** [comment]
So I have been poking a bit at this, and do not have a definitive answer yet. I discovered that the example was crashing trying to free a bit of memory that was not allocated, in `CoreMidiInputPort.cpp` at line 261, `free((void *) memoryReference);`

Now, I am not going to be nearly as efficient at debugging this as Derek might because I did not create these classes and I don’t understand the details of their relationships yet, but I did notice a few things that made me go “hmm.” The `open()` and `close()` methods in `CoreMidiSource` do not protect themselves against being called more than once, and in those circumstances will try to free the same block of memory more than one time, which would lead to this kind of crash.

Indeed, I just tried adding protection against this, and the crash has gone away, the example program has been running with no problems for twenty minutes now. I will want to go through the other classes and add similar protections, but this seems a step in the right direction. In the mean time I will make a preview release with this fix in it and see if it addresses Sean’s ( @eclab ) issue at all.

@DerekCook I also notice that the `close()` method in `CoreMidiSource` simply clears the list of transmitters; should it not also close them?

**Comment by brunchboy** [comment]
OK, @eclab here is a preview version with this fix present, can you see if it behaves better for you? I have to zip it up, as we noted earlier in the discussion.

> I have removed this earlier snapshot because it has been superseded by a newer one below.

**Comment by brunchboy** [comment]
@DerekCook I fixed a couple of places where we were returning `NULL` when we should have been returning an empty list according to the SPI spec as well. But I also see that we are not maintaining a list of receivers in `CoreMidiDestination` (the comment says “We do not maintain a list of receivers, as they tend to be transitory context”). To properly implement the MIDI SPI spec, we should probably go ahead and add this feature. We are also supposed to remove receivers and transmitters from our lists when they are closed. Doesn’t CoreMidi keep these lists for us already, perhaps?

**Comment by brunchboy** [comment]
Everyone, here is a newer snapshot in which I have gone ahead and implemented the rest of the things we are supposed to do in terms of tracking transmitters and receivers that we create, as well as removing them when they get closed, and closing them when their parent MIDI device closes. This makes us much more correctly implement the MIDI Service Provider Interface contract, and will hopefully fix the problems @eclab was seeing. @DerekCook , this is enough of a change that you probably want to test it with some of your synths as well. But if it looks good to everyone, we should probably make another release with it soon.

> Again, snapshot removed because there is a newer one below.

**Comment by brunchboy** [comment]
And here is one more (last, I hope?) snapshot version to try. I further noticed that we were not closing our devices when we notice them disappearing from the MIDI environment. This cleans that up, so that if anyone is holding on to a device reference for something that has been detached from the MIDI environment (or any receivers or transmitters associated with such a device), they will properly be marked as closed, and give you appropriate exceptions when you try to use them at that point. Again, you will have to unzip this in order to test it. 
[coremidi4j-1.1-SNAPSHOT.jar.zip](https://github.com/DerekCook/CoreMidi4J/files/1118113/coremidi4j-1.1-SNAPSHOT.jar.zip)

**Comment by eclab** [comment]
All, I'm going to have exceptionally poor internet access for about a month due to vacation location.  I'll try to do a bit of testing but the turnaround could be weeks.  Thanks for getting on this.

Sean

**Comment by brunchboy** [comment]
No worries! Our turnaround was not particularly fast in responding to your report, so that is only fair. 😁 And I suspect it may be a while before Derek has time to do his testing too.

**Comment by DerekCook** [comment]
James, Thanks very much for getting here first. I of course trust your judgement in what the correct behaviour should be, and the focus in the early days was getting something that worked that undoubtedly would need improvement! 

I will be back home soon, and will be doing a new release of my librarians as a have quite a few fixes batched up (have been able to work on them whilst away on a laptop (but no synths and MIDI devices), so will need to do some synth testing, and I will do that with then new release of CoreMIDI4J.

**Comment by brunchboy** [comment]
That sounds perfect. And of course, I understand well the challenge of getting this library working at all, and the plan of polishing rough edges as we find them! I was so daunted by the prospect of the initial implementation that I waited for you to come along and do it. I’m glad to be able to contribute a bit from time to time.

**Comment by DerekCook** [comment]
Have tried the new release with all of my librarian and synth combinations, and all looks good. But suggest we keep this open until we hear from Sean.

**Comment by brunchboy** [comment]
Thanks, that’s great (and fast)! And I agree.

**Comment by brunchboy** [comment]
@eclab I hope you had a lovely vacation, and will have a chance to test this fix.

**Comment by eclab** [comment]
Not back yet!  Back in a week.  So you have to give me a few.  :-)

**Comment by DerekCook** [comment]
Hi, when you get time, please give this a go and let us know how it works for you.

**Comment by eclab** [comment]
I'm ready to test the new version of the code: but it looks like there's no jar file available, I'd have to build it from source (which I'm not set up to do).  Can someone build a jar of the current code, meant for handling issue #19, and stick it somewhere I can grab?

**Comment by brunchboy** [comment]
If you scroll up a bit you will find it in this thread, attached to my comment on July 2.

**Comment by eclab** [comment]
Aaaaand I am a big dummy.  Will get on it.

Sean

**Comment by brunchboy** [comment]
Not at all! There was a lot of other stuff discussed since then, it was very easy to miss.

**Comment by eclab** [comment]
I am happy to report that I cannot reproduce the errors at this point using the build snapshot above.  I have tested both with the test code I submitted earlier and with production code in Edisyn.

**Comment by brunchboy** [comment]
That’s excellent news, thanks so much for testing and reporting that. Derek, should I make a release with these fixes, or should we wait until you have a chance to finish the other enhancements you are working on as well?

**Comment by DerekCook** [comment]
Hi, James. 

I think I am done with the changes I plan to do on #21

Haven't looked at #22 yet (I really need to get back to some other things for a while), and one of the reasons for using your Maven build was to see if I got the problem in #20, which I didn't

So I think a combined release that gets #19 and #21, once others have had a chance to test, would make sense.

**Comment by eclab** [comment]
Any ETA on when a release might happen?  I'd like to release a new version of Edisyn, but I'd prefer to wait so I can release it with the new version of CoreMIDI4J.

**Comment by brunchboy** [comment]
I could push a release tonight if Derek feels it is ready. Has anyone had a chance to test the snapshot that Derek posted in the discussion for issue #21? @DerekCook, there haven’t been any changes since then, have there?

**Comment by DerekCook** [comment]
I'm happy if people are happy with their own testing after #21 and the snapshot on it, which provides new device names - it works for me. I have done no changes since my last commit, so the Master Branch is up to date with my changes for #19 and #21

**Comment by brunchboy** [comment]
All right, I will prep it for release, and try Beat Link Trigger with a local copy. I will also update a few of the Maven plugins we use, which have had releases this past week, and if all goes well, will release it. @eclab did you have a chance to test the build in the issue #21 discussion?

**Comment by brunchboy** [comment]
All right, I definitely like the way you can now see the user-assigned device device names, as well as the combined device and port names when there are multiple ports for a device. One thing that is a little disconcerting, although I don’t know if there is anything we can do about it, is that changing device names in Audio MIDI Setup does not pass those changes through to Java until the device is disconnected and reconnected, or the JVM is shut down and relaunched. There may just not be an event to notify us of such changes? This is particularly confusing in the case where I start with just one port defined in the IAC Driver (“Bus 1”), which shows up in CoreMidi4J with the name “IAC Driver”. If I then add another port, “Bus 2”,  CoreMIDI4J shows that I have two devices "IAC Driver” and “IAC Driver Bus 2”. Both of them are selectable and work. But if I then take the IAC Driver device offline and back online, or quit and restart the JVM, suddenly the names are, more sensibly “IAC Driver Bus 1” and “IAC Driver Bus 2”.

Since I save my port connections by name, that means that a configuration that works in one run will break the next time I run the program, simply because I added a second port to the IAC driver, which is likely to confuse me or my users.

Looking a bit deeper, I see the following comment in `CoreMidiClient.cpp` for `notifyCallback`:

```cpp
// Message IDs (enum MIDINotificationMessageID)
//      1 kMIDIMsgSetupChanged
//      2 kMIDIMsgObjectAdded
//      3 kMIDIMsgObjectRemoved
//      4 kMIDIMsgPropertyChanged
//
//  In experimenting with notification of changes, you get four messages, in the sequence 4, 3, 3, 1 when removing an interface and 4, 2, 2, 1 when adding an interface
//  I am guessing that the duplicate 3, 3 or 2, 2 is because the deves in question have a source and a destination.
//
//  I think we only need to react to one message, the final one, not all four, assuming that we will use this as a trigger to rescan the interfaces.
```

And I think this means we can fix this problem. Now that we are affected by property changes (the renaming of devices), we should let Java know when we receive those as well as when we receive setup changes. I will try making that change.

**Comment by brunchboy** [comment]
Well, unfortunately it’s not that simple. Trying to react to both `kMIDIMsgSetupChanged` and `kMIDIMsgPropertyChanged` led to `java.util.ConcurrentModificationException` being thrown several times in the `notifyCallback` method, presumably because we are trying to handle additional MIDI changes while still processing the first. I am going to try a slightly different approach. If we get a second callback while already in the process of handling one, I am simply going to set a flag saying we need to re-run our handler when it finishes. That should let us batch multiple callbacks into a single cleanup pass at the end, and also can probably let us stop trying to discriminate on what specific kind of change has occurred, which will probably be safer anyway.

**Comment by DerekCook** [comment]
Good spot. Proposed change makes sense to me.

**Comment by brunchboy** [comment]
Unfortunately it turns out to be a deeper problem, and as it's well past my bed time already, I won't be finished tonight. And as Internet service is down in Madison as well now, this is only a quick summary from my phone. The concurrency issue is a bug resulting from client code in my application changing the listener list on another thread in response to the first environment change. I want to fix that by changing from an `ArrayList` to a more modern and safe `ConcurrentHashmap` which will also protect against double adds. That with the change mentioned above will eliminate the need to synchronize the environment change callback from CoreMidi, which will make us better citizens. 

But the big problem is that the current code assumes that if the unique identifier is already known, the device is unchanged. That is clearly no longer true. I think we are going to have to make the `CoreMidiDeviceInfo` member no longer `final`, and reload it whenever the environment changes. Derek, does that sound correct and sufficient and safe to you?

So we are a bit out from a release yet. I probably won’t have time to finish this until the weekend.

**Comment by DerekCook** [comment]
Hi, James. I'm not clear as to why the UUID changes? I trust your judgement of course, and in my librarian use case I am not terribly interested in devices coming and going. So it is of course more important that you get it working as needed for your use case, and then mine should be fine.

If it helps, there is no reason that we (assuming it is not difficult) cannot get a release for #19 out of the way and do another for #21 when ready if splitting them suits everybody. I was just trying to save the pain of doing two releases in a short period if it was all easy :)

**Comment by brunchboy** [comment]
Sorry, Derek, the problem is not that the unique ID changes, but that it _doesn’t_ change. As I feared, my phone-based brevity led to confusion. In `CoreMidiDeviceProvider`’s `buildDeviceMap()` method, we have:

```java
// If the unique ID of the end point is not in the map then create a
// CoreMidiSource object and add it to the map
if ( !midiProperties.deviceMap.containsKey(uniqueID) ) {
        midiProperties.deviceMap.put(uniqueID,
          new CoreMidiSource(getMidiDeviceInfo(endPointReference)));
}
```

However, in the case where a user has edited a device name, that device’s `uniqueID` is already present in the device map, so we are assuming we already know everything we need to about it, and do not update the `CoreMidiDeviceInfo` object. So the changed name is not reflected in the values that we are reporting to our clients. Does that make more sense as an explanation?

**Comment by brunchboy** [comment]
So what I am saying is that I believe we need to act differently for devices that are already present in `midiProperties.deviceMap`: We don’t want to replace the `CoreMidiSource` object, since we may be tracking opened connections to it, but we _do_ want to update it with a new `CoreMidiDeviceInfo` object containing up-to-date name information in case that has changed. And that is what I was asking whether it would be safe and sufficient from your perspective.

**Comment by brunchboy** [comment]
I would have no idea about how to tease out the changes for the two separate issue fixes into two releases, and that seems like much more work to me than simply finishing what we have started. But if you prefer to take that route, I can abandon the work I have done so far, and try again once you’ve released what you want to.

**Comment by DerekCook** [comment]
Hi, James

Ok I'm with you. No, do not abandon anything. It can wait as far as I'm concerned. Path of least resistance rules. 😀

**Comment by DerekCook** [comment]
Ps not sure if we can make CoreMidiDeviceInfo to not be final, as, from memory, MidiDevice.Info is final. But I may be barking mad and it has been a very long week!!

**Comment by brunchboy** [comment]
You are not mad; or at least, you are not providing evidence of that in this discussion. 😉 I have had a bit of time to investigate this, and even though Java’s `MidiDevice.Info` is final, our own `CoreMidiSource` and `CoreMidiDestination` classes have their own `info` members which we can make non-final. I hope to have an updated snapshot for the two of you to test shortly that will reflect updated device names. I am also fixing another issue I discovered: Java is constructing `CoreMidiDeviceProvider` over a hundred times, and each time we are adding it to the listener list to be notified of MIDI environment changes, so we are doing *far* more work than we need to, rebuilding the device map over and over, when the environment changes. I am going to prevent more than one instance being added to the listener list, because one update of the (shared) map will suffice.

**Comment by brunchboy** [comment]
OK, things are working nicely and I am in the home stretch now. It turns out that even with a simple name change, we do still receive the final `kMIDIMsgSetupChanged` message. So we don’t need to change which messages we react to after all, and can simply rescan the interfaces once upon receiving that. Once I get a bit more testing done, I will upload a snapshot. I hope that you both can do some testing with it, because of the variety of changes, but if it looks good we should be able to release it.

**Comment by brunchboy** [comment]
OK, here is a new snapshot for you to test! @DerekCook and @eclab if this works well for you I will promote it to a release. (Remember to unzip the file before trying to use it.)
[coremidi4j-1.1-SNAPSHOT.jar.zip](https://github.com/DerekCook/CoreMidi4J/files/1283040/coremidi4j-1.1-SNAPSHOT.jar.zip)

**Comment by DerekCook** [comment]
Thanks, James

Am away on business right now, so will check this out when I am back home for the weekend.

**Comment by brunchboy** [comment]
Thinking about it, I have a hard time imagining any of my changes this week affecting your librarians, so if @eclab and @odbuser2 are happy with the above snapshot I could go ahead and release it, and if by some amazing bad luck it did cause problems for you I can fix that later. Or we could just wait for the weekend, depending on how much of a hurry @eclab is in to release Edisyn.

**Comment by eclab** [comment]
I can wait, it's fine.  But I've not been able to test the new snapshot, I'll try to do that tonight.

Sean

**Comment by eclab** [comment]
The new snapshot bombs on the example I had provided earlier (using OS X 10.9.5, java 1.8.0_131).  The snapshot prior to that does *not* bomb, so perhaps we're seeing a regression.

I'm attaching the bomb message and the associated log file.

    # A fatal error has been detected by the Java Runtime Environment:
    #
    #  SIGSEGV (0xb) at pc=0x00007fff9329ca1f, pid=1130, tid=0x0000000000000d07
    #
    # JRE version: Java(TM) SE Runtime Environment (8.0_131-b11) (build 1.8.0_131-b11)
    # Java VM: Java HotSpot(TM) 64-Bit Server VM (25.131-b11 mixed mode bsd-amd64 compressed oops)
    # Problematic frame:
    # C  [CoreFoundation+0x12a1f]  CFStringGetLength+0xf
    #
    # Failed to write core dump. Core dumps have been disabled. To enable core dumping, try "ulimit -c unlimited" before starting Java again
    #
    # An error report file with more information is saved as:
    # /private/tmp/hs_err_pid1130.log
    #
    # If you would like to submit a bug report, please visit:
    #   http://bugreport.java.com/bugreport/crash.jsp
    # The crash happened outside the Java Virtual Machine in native code.
    # See problematic frame for where to report the bug.

[hs_err_pid1130.log.zip](https://github.com/DerekCook/CoreMidi4J/files/1286505/hs_err_pid1130.log.zip)

**Comment by brunchboy** [comment]
Well, that’s frustrating, especially since I did not change any native code in this last round of fixes. Did you try the latest snapshot Derek posted which  includes the device name disambiguation fixes on issue #21? And this is why I wanted people to test it before building a release, so thanks for doing that.

And thank you for including the log file. I guess I can try rolling back my changes to rebuild the device map on a separate thread, especially since we don’t need to handle as many types of change notifications as I feared at one point. That is probably the biggest change I made. But I guess I need to know if the snapshot on the issue #21 thread crashes or not. Let’s see if I can find and remember how to run your example.

**Comment by brunchboy** [comment]
Hmm. I just tried running your example for ten minutes and there was no crash. I added a counter to show how many passes through the infinite loop it had reached, and I am up to over 11,000 passes on a second run as I post this, and still no crash. Are you sure you were using the right snapshot in your class path?

**Comment by brunchboy** [comment]
I let it run to 20,094 passes before killing it.

macOS Sierra 10.12.6
java version "1.8.0_121"
Java(TM) SE Runtime Environment (build 1.8.0_121-b13)
Java HotSpot(TM) 64-Bit Server VM (build 25.121-b13, mixed mode)

**Comment by eclab** [comment]
The one I tried was the one in the message that goes "OK, here is a new snapshot for you to test! @DerekCook and @eclab if this works well for you I will promote it to a release. (Remember to unzip the file before trying to use it.)".  Tried it again, bombs on launch.  :-(

The one prior to that (the "And here is one more (last, I hope?) snapshot version to try.") works fine.

**Comment by brunchboy** [comment]
Thanks for the clarifications. The one that works fine does not include any of the changes Derek made to fix issue #21 so could you try the latest snapshot in that thread and let us know what happens?

Unfortunately, since I can’t reproduce the crash here, I am not sure what I am going to be able to do to fix it. You have a far older Mac OS version than I have had available for years.

**Comment by eclab** [comment]
Github really needs uniquely numbered comments.

You want me to try the one in issue #21 that starts with "Reporting in again a little later than intended. Quite a lot going on in my life! But have been chipping away slowly at it."?  Hang on.

**Comment by brunchboy** [comment]
Yes, that looks like the correct one. Thanks! We probably should have done separate releases to address these individual issues, in hindsight.

**Comment by eclab** [comment]
I'm afraid that one also appears to be bombing.

    # A fatal error has been detected by the Java Runtime Environment:
    #
    #  SIGSEGV (0xb) at pc=0x00007fff9329ca1f, pid=23772, tid=0x0000000000000d07
    #
    # JRE version: Java(TM) SE Runtime Environment (8.0_131-b11) (build 1.8.0_131-b11)
    # Java VM: Java HotSpot(TM) 64-Bit Server VM (25.131-b11 mixed mode bsd-amd64 compressed oops)
    # Problematic frame:
    # C  [CoreFoundation+0x12a1f]  CFStringGetLength+0xf
    #
    # Failed to write core dump. Core dumps have been disabled. To enable core dumping, try "ulimit -c unlimited" before starting Java again
    #
    # An error report file with more information is saved as:
    # /private/tmp/hs_err_pid23772.log
    #
    # If you would like to submit a bug report, please visit:
    #   http://bugreport.java.com/bugreport/crash.jsp
    # The crash happened outside the Java Virtual Machine in native code.
    # See problematic frame for where to report the bug.

[hs_err_pid23772.log.zip](https://github.com/DerekCook/CoreMidi4J/files/1286632/hs_err_pid23772.log.zip)

**Comment by brunchboy** [comment]
Thanks again for following up. OK, that means it does not make sense for me to try to examine the changes I made this week; it was caused by the original device name fixes. And what you say about it crashing on launch is consistent with that stack trace, it appears to happen when the `CoreMidiDeviceProvider` is instantiated and builds the original device map. I wonder if Derek will be able to reproduce it either? If not, we may need to make a branch that is based on my working snapshot, which fixes only this issue, and release that for you, and spend longer trying to see if we can fix the #21 fixes in a way that doesn’t introduce this regression for you, otherwise you will be stuck on that release. But at least you will have something.

**Comment by brunchboy** [comment]
Looking at the native code implicated in the crash, it seems there are only a few lines where the problem might be:

```objc
  if ( buildName != NULL ) {

    CFStringAppend(buildName, deviceName);
    
    if ( numberOfEntities > 1 ) {
      
      CFStringAppend(buildName, CFSTR(" "));
      CFStringAppend(buildName, entityName);
      
    }
    
    // Overwrite the deviceName with our updated one
    deviceInfoName = CFStringCreateCopy(NULL, buildName);

    // And release the temporary string
    CFRelease(buildName);

  }
```

The crash happens in a call to `CFStringAppend`, so I would say that either the value of `deviceName` or `entityName` is `NULL` or otherwise invalid to cause the crash. I have just created another snapshot in which I have uncommented a bunch of debugging lines that Derek put in the code right before this section, which should print out all the values we have loaded from CoreMidi, including those two variables (as `Device kMIDIPropertyname` and `Entity kMidiPropertyName` respectively). It would be interesting to see what values are printed out right before this crashes for you.
[coremidi4j-1.1-SNAPSHOT.jar.zip](https://github.com/DerekCook/CoreMidi4J/files/1286689/coremidi4j-1.1-SNAPSHOT.jar.zip)

**Comment by eclab** [comment]
BTW, I know you're working on naming schemes for MIDI devices.  I notice that in the first snapshot you put out (the one in July or so, the one that's working on my box) it seems to be setting getDescription() and getName() to return the same thing.  I'm not sure if this was intentional, or even a bug.  But it seems that the description should be more descriptive than the name.

**Comment by brunchboy** [comment]
I will let Derek respond about naming questions since I didn’t pay close attention to what he was able to get there, although you will see all the information that is available if you run the new snapshot in the comment you were responding to. I am very eager to find out what gets printed right before your crash takes place. I think it is due to some difference in your MIDI environment, probably a null value we are not expecting. Do you think you will be able to try that and share the results?

**Comment by DerekCook** [comment]
Yes, the whole reason for working on the naming was to remove issues like that one. 

I agree with James: please run the debug version that James has provided (either in your Java IDE or from the terminal so we can see the debug info. Will be interesting to see what your ENV provides. I can add some more error checking defensive programming to stop null pointers, but it will be better to see what is causing the problem if you can catch it (it is not happening here).

**Comment by DerekCook** [comment]
Hi,

I have pulled the latest updates and built the library here and it is all working fine with my an.factory librarian (for the Yamaha AN1x, which is the lighter synth to pull from my keyboard room to where my MAC is!).

So I am guessing it must be a specific environment issue.

I then enabled the debug info I put into the `getMidiDeviceInfo()` native method (what James has done on the snapshot above). After verifying it is working, I then added some test code to deliberately set one of the string pointers to NULL to see what happened, and I immediately got a crash

> CoreMIDI4J Java classes are available.
> Working MIDI Devices:
> End Point 
>   End Point Reference                  129155114
>   End Point kMIDIPropertyName          Mobile Keys 25
>   End Point kMIDIPropertyModel         Mobile Keys 25
> 
>  A fatal error has been detected by the Java Runtime Environment:
> 
>   SIGSEGV (0xb) at pc=0x00007fff923d8b52, pid=31531, tid=7171
> 
>  JRE version: Java(TM) SE Runtime Environment (8.0_51-b16) (build 1.8.0_51-b16)
>  Java VM: Java HotSpot(TM) 64-Bit Server VM (25.51-b03 mixed mode bsd-amd64 compressed oops)
>  Problematic frame:
>  C  [libsystem_c.dylib+0x1b52]  strlen+0x12
> 
>  Failed to write core dump. Core dumps have been disabled. To enable core dumping, try "ulimit -c unlimited" before starting Java again
> 
>  An error report file with more information is saved as:
>  /Development/CoreMidi4JTest/hs_err_pid31531.log
> 
>  If you would like to submit a bug report, please visit:
>    http://bugreport.java.com/bugreport/crash.jsp
>  The crash happened outside the Java Virtual Machine in native code.
>  See problematic frame for where to report the bug.
> 
>   **End Point kMIDIPropertyManufacturer**  

And it was the `endpointManufacturer` variable that I set to NULL

So if you run James's snapshot above (5 posts up, including this one), you will still get the crash, at the first field which is NULL, but if multiple fields are NULL, then that will take a long time to diagnose. I will therefore update the code to provide some defensive programming, including the debugging strings. Watch this space.....

**Comment by DerekCook** [comment]
OK. Attached to this reply is a SNAPSHOT JAR that has **ALL** of the latest fixes, and I have also hopefully added in debug/test code in `getMidiDeviceInfo()` native method to stop the JVM crashes. If any name is a NULL pointer then it will be replaced with an error message and you can see that in the name being returned. Will be interesting to see what is NULL in your environment.

It also still contains all of the debug trace uncommented so we can precisely see what the OS X property calls are returning.

So this is not the final solution, but it should stop the crashes and gather debug info for us. Please run as soon as you can and let us know the outcome. 

[coremidi4j-1.1-SNAPSHOT.jar.zip](https://github.com/DerekCook/CoreMidi4J/files/1290524/coremidi4j-1.1-SNAPSHOT.jar.zip)

**Comment by eclab** [comment]
Log attached:

[log.zip](https://github.com/DerekCook/CoreMidi4J/files/1291115/log.zip)

This is a log of the following activities while running Edisyn:

1. Fire up Edisyn when nothing is attached.
2. Run MIDI Monitor (a software program which creates a USB virtual MIDI port).
3. Run Sysex Librarian  (a software program which creates a USB virtual MIDI port).
4. Attach a PreenFM2 synth.
5. Disconnect the Preen and attach to a Novation Zero Mk ii.  (ports 1 and 2 of a generic MIDI device it seems)
6. Disconnect the Zero and attach a Tascam US 2x2 USB MIDI/Audio input device (also lists itself as a generic MIDI device)
7. Disconnect the US 2x2 and attach a Korg Microsampler (creates four ports)
8. Disconnect the Korg and fire up MIDI Monitor and Sysex Librarian again

This seem to be working more smoothly.  However there's a bug when attaching MIDI Monitor and Sysex Librarian.  Edisyn displays names of devices in a pop-up menu by using the following code:

    String desc = device.getDeviceInfo().getDescription().trim();
    // All CoreMIDI4J names begin with "CoreMIDI4J - ", so that's substringed out
    String name = device.getDeviceInfo().getName().substring(13).trim();
    if (name.equals(""))
        return desc; 
    else 
        return name;

Using this code, when the software devices are attached, the resulting string displayed is

    "** NULL POINTER for Device Name **"

**Comment by brunchboy** [comment]
Thanks so much! This is hugely valuable information, and it looks like the crash was exactly what we thought it would be, and Derek’s latest round of defensive changes have in fact fixed it.

Derek, I have been able to reproduce the crash in the old code, and the output that Sean reported here using your latest fixed/debug build, by launching MIDI Monitor and checking “Act as a destination for other programs” in the Sources list. So evidently certain kinds of software destinations have `null` values for Name, Model, and Manufacturer.

Since those can happen, perhaps rather than `** NULL POINTER for Device Name **` etc., we should simply return something like `<No Name>`, `<No Model>`, `<No Manufacturer>`? Or we could just leave things as they are, turn off the debugging print statements again, and release this. What do you guys think?

**Comment by brunchboy** [comment]
Also, it looks like that SysEx Librarian program that causes the same problem is similar to Derek’s own x.factory Librarians, which was the reason he wrote CoreMidi4J in the first place. 😄

**Comment by brunchboy** [comment]
Although I guess I don’t like what we are getting for entity and device from MIDI Monitor, in particular the fact that the supposedly unique references are zero and not unique. Perhaps we should just skip/ignore those entries entirely?

```
End Point 
  End Point Reference                  4721806
  End Point kMIDIPropertyName          MIDI Monitor (Untitled)
  End Point kMIDIPropertyModel         MIDI Monitor
  End Point kMIDIPropertyManufacturer  Snoize
  End Point kMIDIPropertyOffline       0
  End Point kMIDIPropertyUniqueID      245106699
  End Point kMIDIPropertyDriverVersion 0

  Entity    Reference                  0
  Entity    kMIDIPropertyName          ** NULL POINTER **
  Entity    kMIDIPropertyModel         ** NULL POINTER **
  Entity    kMIDIPropertyManufacturer  ** NULL POINTER **
  Entity    kMIDIPropertyOffline       0
  Entity    kMIDIPropertyUniqueID      0
  Entity    kMIDIPropertyDriverVersion 0

  Device    Reference                  0
  Device    kMIDIPropertyName          ** NULL POINTER **
  Device    kMIDIPropertyModel         ** NULL POINTER **
  Device    kMIDIPropertyManufacturer  ** NULL POINTER **
  Device    kMIDIPropertyOffline       0
  Device    kMIDIPropertyUniqueID      0
  Device    kMIDIPropertyDriverVersion 0

  Number of entities                   0
```

**Comment by brunchboy** [comment]
Hmm… and while Sean’s crash test program is running, if I uncheck “Act as destination for other programs” in MIDI Monitor, I get the following output:

```
java.lang.IllegalArgumentException: Requested device not installed: CoreMIDI4J - ** NULL POINTER for Device Name **
	at javax.sound.midi.MidiSystem.getMidiDevice(MidiSystem.java:226)
	at Example.gatherMidiDevices(Example.java:190)
	at Example.main(Example.java:303)
```

Looking at his code, that’s probably just an artifact of the fact that the test program is busy-looping over all the MIDI devices continuously, rather than responding to notifications about changes in the MIDI environment, so when the device disappears between asking for the list of all devices, and trying to ask for it, we see that exception. I do not think this is a problem we need to worry about.

**Comment by eclab** [comment]
On Sep 10, 2017, at 10:25 PM, James Elliott <notifications@github.com> wrote:

> Since those can happen, perhaps rather than ** NULL POINTER for Device Name** etc., we should simply return something like <No Name>, <No Model>, <No Manufacturer>? Or we could just leave things as they are, turn off the debugging print statements again, and release this. What do you guys think?

Well, using your original code, Edisyn would display "Sysex Librarian" and "MIDI Monitor".  That was either in the name or in the description.

So I'm somewhat confused here.  It would seem to me that the right thing to return as the MIDI device name would be the EndPoint kMIDIPropertyName, and perhaps for the MIDI description, the EndPoint kMIDIPropertyModel.

But it sounds like you're suggesting using the Entity's kMIDIPropertyName rather than the EndPoint's kMIDIPropertyName (and likewise for the kMIDIPropertyModel).  Use of this was what was triggering the bug, since the Entity's name and description are null.

Furthermore, using the Entity's version, you'd get wrong values ("Port 2" rather than "KBD/KNOB" for example) for the MidroSampler.  

Sean

**Comment by brunchboy** [comment]
Right, now we are getting into details that Derek will be best suited to answer, since he made these changes. That makes sense to me too. But it looks like the current code is doing the opposite. It is using the device name, but then if there is more than one entity for the endpoint, it appends the endpoint name.

Derek, perhaps we should swap that? Use the endpoint name always, and if there is more than one entity append the device name? It looks like that will go back to displaying what Sean is used to seeing here, because it seems that in the case of MIDI Monitor at least, the number of entities is coming back as zero, so we would simply be ignoring these null device names.

And I think that would make life easier for my users as well, because it will mean that the MIDI outputs configured in their existing triggers are not going to change; I was getting ready to explain to people that after the next release of Beat Link Trigger they were going to have to go through and edit all their triggers, and select the correct new name of the MIDI output they wanted the trigger to send to.

But perhaps there is an important reason you want it to work this way? If so, I guess an alternative approach would be to fall back to the endpoint name when the device name is `null`?

**Comment by DerekCook** [comment]
Thanks, both. Quite a lot to digest at 6 in the morning, but the debug code has shown us what is causing the issue, so result! 👍 

@eclab  , perhaps I am being a little pedantic, but the return value `"** NULL POINTER for Device Name **" `is not a bug at this point; it's a debugging value as I did not get the crash in my environment, so did not know the cause. Now we are getting the clear indication of the problem, we can address it. 👍 

James, that's useful info on using Snoize Midi Monitor to replicate the issue. I will use that tonight when I do the fix. So it looks like "virtual end points" have no device or entity.

So the fix is quite simple. When you have a device name, use it. When it is NULL use the endpoint name. If both are NULL return `<Unknown Device>` to ensure we do not crash. I will do that tonight and post a new snapshot.

In terms of the question as to "why not revert to the end point name as per the original code", I won't reproduce the detail on #21 (it's there to read) but the reason for doing this change was to allow the user to set unique device names if they have multiple MIDI devices of the same type on their system. So, for example, I have two Focusrite Scarlett 2i2 devices (main and spare), but usually only one plugged in. If I had both plugged in, their CoreMidiDeviceInfo structures, based on EndPoint names, would return:

> Scarlett 2i4 USB
> Scarlett 2i4 USB

I.e. you cannot identify which one is which from the EndPoint name (the issue raised in #21), but in the OS X MIDI Panel you an edit the Device Description and this is what is returned as the **Device name**. E.g. Here is what I get back via the debug info on my 2i4 (with edited description):

> End Point 
>   End Point Reference                  136466478
>   End Point kMIDIPropertyName          **Scarlett 2i4 USB**
>   End Point kMIDIPropertyModel         Scarlett 2i4 USB
>   End Point kMIDIPropertyManufacturer  Focusrite
>   End Point kMIDIPropertyOffline       0
>   End Point kMIDIPropertyUniqueID      -7005973
>   End Point kMIDIPropertyDriverVersion 0
> 
>   Entity    Reference                  136466477
>   Entity    kMIDIPropertyName          **Scarlett 2i4 USB**
>   Entity    kMIDIPropertyModel         Scarlett 2i4 USB
>   Entity    kMIDIPropertyManufacturer  Focusrite
>   Entity    kMIDIPropertyOffline       0
>   Entity    kMIDIPropertyUniqueID      1755297388
>   Entity    kMIDIPropertyDriverVersion 0
> 
>   Device    Reference                  136466476
>   Device    kMIDIPropertyName          **Cookie's Scarlett 2i4 USB**
>   Device    kMIDIPropertyModel         Scarlett 2i4 USB
>   Device    kMIDIPropertyManufacturer  Focusrite
>   Device    kMIDIPropertyOffline       0
>   Device    kMIDIPropertyUniqueID      226424688
>   Device    kMIDIPropertyDriverVersion 0

I.e. look at the differences in the `kMIDIPropertyName` being returned for the End Point and Device, the latter is my edited description entered in the Device Panel

The other issue that I fixed at the same time is, for example, my Motif Rack ES Synth has 8 end points so returned `Port 1, Port 2, ..., Port 8` as end points with no reference to the Device name. Thanks to the info provided by the OP on #21 I learnt how to work backwards from the end point to get the device name which I can add to the end point name for a device with multiple entities/endpoints, which fixes both issues (but caused this new problem).

Hope that all makes sense. To conclude, for reasons given above, it is not logical to reverse the #21 fix, and the fix for the introduced issue is simple and it won't take long: If Device Name is NULL fall back to using the Endpoint Name, if both are NULL return an `<Unknown Device>` value.

**Comment by eclab** [comment]
On Sep 11, 2017, at 1:36 AM, Derek Cook <notifications@github.com> wrote:

> So the fix is quite simple. When you have a device name, use it. When it is NULL use the endpoint name. If both are NULL return <Unknown Device> to ensure we do not crash. I will do that tonight and post a new snapshot

In the microSAMPLER case (and probably microKORG etc.), all three names (endpoint, entity, device) are filled out, but only one of them is correct: the endpoint name.  I think the proposed fix will screw up these owners.  Certainly it's gonna hurt me both as a developer and as the owner of these things.  :-(

> In terms of the question as to "why not revert to the end point name as per the original code", I won't reproduce the detail on #21 (it's there to read) but the reason for doing this change was to allow the user to set unique device names if you have multiple MIDI devices of the same type on your system.

The problem with this is that it's potentially hurting the 90% userbase, who now all have to root through the MIDI Control panel, and the developers, who now have to field irritated questions from users who don't know why their device names are weird, just help out the 10% userbase (people with multiple copies of the same device).  Spock would definitely not approve.

I think the situation you have is:

1. A few people need to distinguish multiple instances
2. Some devices have a null device name
3. Many devices require the entity name to be used in order to be correct
3. Many devices have the *same* name for the device and the entity
4. We don't want a great many non-technical people to have to understand how to to the mysterious MIDI control panel to change every one of their devices to avoid doubling of the name or having to change it to the correct name just to serve people in #1.
5. Lots of devices just say "USB MIDI Device".  We want to enable people to be able to change that name.

Let me suggest an alternative.  If both names are empty, use "USB MIDI Device".  If the device name is empty or it's the same as the entity name, use the entity name.  If the entity name is empty, or it's the generic "USB MIDI Device", use the device name. Otherwise, use the form "entityname (devicename)" to permit distinguishing.

Code along these lines:

NAME:
    dname = ""
    ename = ""
    If endpoint.kMIDIPropertyName != null && !device.endpoint.kMIDIPropertyName.equals("USB MIDI Device")
        ename = endpoint.kMIDIPropertyName;
    If device.kMIDIPropertyName != null && !device.kMIDIPropertyName.equals(endpoint.kMIDIPropertyName)
        dname = device.kMIDIPropertyName;
    if (ename.equals("") && dname.equals(""))
        return "USB MIDI Device"
    else if (ename.equals(""))
        return dname;
    else if (dname.equals(""))
        return ename;
    else
        return ename + " (" + dname + ")";

This approach works well with all six devices I had presented you in the log.  Namely, the name would be:

1. microsampler:       KBD/KNOB (Port 2)
2. PreenFM2:        PreenFM2
3. MIDI Monitor:    MIDI Monitor (Untitled)
4. Sysex Librarian:     Sysex Librarian
5. Tascam US 2x2:    USB MIDI Device     [Editable to "Tascam US 2X2"]
6. Novation Zero:    Port 1        [Edible to "Port 1 (Novation Zero)"]

This seems much better.

BTW, you might as well also handle the vendor and description.  I'd do them simply:

VENDOR:
    vend = ""
    If endpoint.kMIDIPropertyManufacturer != null
        vend = endpoint.kMIDIPropertyManufacturer
    else if device.kMIDIPropertyManufacturer != null
        vend = device.kMIDIPropertyManufacturer

DESCRIPTION:
    devid = device.kMIDIPropertyUniqueID     // can this be null?  If so, "unknown"
    entid = entity.kMIDIPropertyUniqueID     // can this be null?  If so, "unknown"
    endpointid = endpoint.kMIDIPropertyUniqueID     // can this be null?  If so, "unknown"
    desc = name + ", by " + vend + ", device=" + devid + ", entity=" + entid + ", endpoint=" + endpointid

TOSTRING:
    return "MIDI Device " + desc

**Comment by DerekCook** [comment]
Hi, @eclab 

Based on your log file (a few posts back) I am not sure I fully see your concern about the name change introduced under #21 . 

Unless I am mistaken (using your log examples, and this will need checking) under the new scheme (including the new fix to revert to the endpoint name if the device name is NULL) you should get:

`PreenFM mk2`
`microSampler - MIDI IN`
`microSampler - KBD/KNOB`
`microSampler - MIDI OUT`
`microSampler - Port 1`
`microSampler - SOUND`
`USB MIDI Device - PORT 1` (for the device with two entities)
`USB MIDI Device - PORT 2` (for the device with two entities)
`USB MIDI Device` (for the device with one entity)
`MIDI MONITOR (Untitled)`
`SysEx Librarian`

Looking at the microsampler and USB MIDI Device with two entities, then the above should be more useful.

I am surprised that you think it is beyond the intelligence of the average user to go to the MIDI panel to edit a device name if needed (e.g. to change those confusing multiple USB MIDI Device descriptions if they are different devices (they have different UIDs)). That is hardly rocket science - we are expecting Users to be able to install Java after all to run our applications (esp now Apple do not include Java 6 as standard).

I now have a dichotomy: The OP of #21 says that the current behaviour is wrong, but I am not fully in agreement with what the OP for #21 was suggesting either. You say the revised proposed behaviour is wrong, but don't agree with that either, but I also now have a view of how it should work, which I have implemented (with new fix). Since #21 was raised, I was no longer happy with the status quo either.

Actually I have pre-empted this, but have not explained it yet, so apologies for that. As part of the update to the`getMidiDeviceInfo()` native method I decided, as I would never satisfy all people at the same time if making such a change, to extend the `CoreMidiDeviceInfo` constructor signature to capture all of the info of relevance available via the OSX property calls. So if you are not happy with the behaviour of the strings returned by CoreMIDI4J `CoreMidiDeviceInfo.getName()` method also the `CoreMidiDeviceInfo.toString()` method (both inherited from `MidiDevice.Info` as **final** methods), then you can get the Device Name, Entity Name and End Point names via new getter methods and roll your own name to suit.

Here is a snippet from my debug program which shows how use the new methods.

```
        for (javax.sound.midi.MidiDevice.Info device : midiInterfaces ) {
          
          if ( device instanceof CoreMidiDeviceInfo ) {
            
            CoreMidiDeviceInfo coreMidiDevice = (CoreMidiDeviceInfo) device;
            
            System.out.println("  " + device);
            System.out.println("    Device   Name: " + coreMidiDevice.getDeviceName());
            System.out.println("    Entity   Name: " + coreMidiDevice.getEntityName());
            System.out.println("    Endpoint Name: " + coreMidiDevice.getEndPointName() + "\n");
            
          }

      }
```

This gives when run

```
  CoreMIDI4J - Mobile Keys 25         <--- This is the default name)
    Device   Name: Mobile Keys 25
    Entity   Name: Mobile Keys 25
    Endpoint Name: Mobile Keys 25

  CoreMIDI4J - Cookie's Scarlett 2i4 USB
    Device   Name: Cookie's Scarlett 2i4 USB
    Entity   Name: Scarlett 2i4 USB
    Endpoint Name: Scarlett 2i4 USB

  CoreMIDI4J - Mobile Keys 25
    Device   Name: Mobile Keys 25
    Entity   Name: Mobile Keys 25
    Endpoint Name: Mobile Keys 25

  CoreMIDI4J - Cookie's Scarlett 2i4 USB
    Device   Name: Cookie's Scarlett 2i4 USB
    Entity   Name: Scarlett 2i4 USB
    Endpoint Name: Scarlett 2i4 USB

  CoreMIDI4J - MIDI Monitor (Untitled)
    Device   Name: <Unknown Device>
    Entity   Name: MIDI Monitor (Untitled)
    Endpoint Name: MIDI Monitor (Untitled)
```
I am also passing the Device, Entity and Endpoint references and UIDs. This can also be extended to pass over anything else you think is needed if not currently being provided
 
So, give the attached a go and let me know what happens

[coremidi4j-1.1-SNAPSHOT.jar.zip](https://github.com/DerekCook/CoreMidi4J/files/1293886/coremidi4j-1.1-SNAPSHOT.jar.zip)

I have ran this tonight and all looks good on my system, including running Snoize MIDI Monitor acting as a destination (as can be seen above).

**Comment by eclab** [comment]
On Sep 11, 2017, at 4:32 PM, Derek Cook <notifications@github.com> wrote:

> Unless I am mistaken (using your log examples, and this will need checking) under the new scheme (including the new fix) you should get:

It is possible that I am misunderstanding the proposed scheme.  But what I read from you was:

>> When you have a device name, use it.... If Device Name is NULL fall back to using the Endpoint Name, if both are NULL return an <Unknown Device> value.

However on the Microsampler, the device name is (from the logs):

  Device    kMIDIPropertyName          microSAMPLER

Indeed this is the case for all of the Microsampler's ports.  But the End Point name is (for example):

  End Point kMIDIPropertyName          KBD/KNOB

Similarly on the Novation, the device name is:

  Device    kMIDIPropertyName          USB MIDI Device

But the End Point name is (for example):

  End Point kMIDIPropertyName          Port 2

I believe the actual outputs under your scheme would be:

Preen:			PreenFM mk2
Microsampler 1:		microSAMPLER
Microsampler 2:		microSAMPLER
Microsampler 3:		microSAMPLER
Microsampler 4:		microSAMPLER
Novation port 1:	USB MIDI Device
Novation port 2:	USB MIDI Device	
Tascam:			USB MIDI Device
MIDI Monitor:		MIDI Monitor (Untitled)
Sysex Librarian:	SysEx Librarian


> I am surprised that you think it is beyond the intelligence of the average user to go to the MIDI panel to edit a device name if needed (e.g. to change that confusing multiple USB MIDI Device if they are different devices (they have different UIDs). That is hardly rocket science - 

I would prefer not to get into opinions of user intelligence here.  However when you're distributing an application you have to think about your whole userbase, and the impact it'd have on your support structure.

> we are expecting Users to be able to install Java after all to run our applications (esp now Apple do not include Java 6 as standard).

I think the generally accepted distribution for public-facing Java applications on the Mac at this point is to do bundles using Oracle's javapackager tool.  This produces a .app which contains all necessary jar files *and* its own copy of a Java VM.  The user does not install Java at all, nor should they be required to.  Instead, they just download your program and double-click on it.


> I now have a dichotomy: The OP of #21 says that the current behaviour is wrong. You say the revised proposed behaviour is wrong. I am more inclined to go with the view I currently have rather than yours (you can tell me I am wrong and show me why), but I am not fully in agreement with what the OP for #21 was suggesting either. So what to do?

I want to make sure I understand this properly.  As I understood it, #21's problem was that if you have multiple copies of the same device, they all show up with the same exact name.  He was suggesting augmenting it with the ID (which is a big number), and I *believe* you were suggesting instead to just use the device name because you can change that uniquely for different devices in the MIDI control panel.  Correct?

If so, I'm pretty sure my approach would fully handle #21's issues, AND handle NULL, AND enable better naming than "USB MIDI Device", AND also maintain proper naming when you don't have multiple instances.  So what's not to like?


> Actually I have pre-empted this, but have not explained it yet, so apologies for that. As part of the update to thegetMidiDeviceInfo() native method I decided, as I would never satisfy all people, to extend the constructor signature to capture all of the info of relevance available via the OSX property calls. So if you are not happy with the behaviour of the strings returned by CoreMIDI4J CoreMidiDeviceInfo.getName() method also the CoreMidiDeviceInfo.toString() method (both inherited from MidiDevice.Info), then you can get the Device Name, Entity Name and End Point names via new getter methods and roll your own name to suit.

This requires that I special case for CoreMidi4J's API for Macs only, rather than relying on Java MIDI.  :-(  Before you did that, please take a look again at my proposal.  I think (?) it handles all the issues brought up, so I'm not sure why you don't like it.  Is it possible you may have misread it?  I do that a lot.

Sean

**Comment by brunchboy** [comment]
It seems like the confusion arising here is whether your microSampler is showing up a single device with multiple ports or as multiple devices. I believe if it is a single device, you will see disambiguated names as described by Derek; if it is showing up as independent devices, then indeed it would be a whole bunch of indistinguishable device names that the user would have to sort out manually by renaming.

I agree with Derek that it is necessary to always show the device name when it is available (that is, not `null`), since that is the only thing that the user is able to edit.

From trying to follow all the snippets and listings and logs, and trying to keep things simple, what about the following approach?

* If either of the endpoint or device name is `null`, return only the other. (If both are `null`, just return some generic “unknown device” name as already proposed).

* If the endpoint and device name are both present, and both have the same value, return just one of them, to avoid redundancy.

* If the endpoint and device name are both present and different, return something like `Device Name (Endpoint Name)` or `Endpoint Name - Device Name`, whatever seems to make the most sense.

That way it is maximally unambiguous, the user can always control an element of it when necessary to eliminate duplicates, and it is not more redundant than it needs to be. Does that sound reasonable?

**Comment by brunchboy** [comment]
And to clarify my previous comment, the reason I am thinking we may want to display both the device and endpoint name even for endpoints with only one entity is that it sounds like Sean is reporting that there are manufacturers whose device name is always the same even for different kinds of devices, and for which it is only the endpoint name that gives the user (without recourse to manual intervention via the Audio MIDI Setup configuration interface) the ability to tell which one is which.

**Comment by eclab** [comment]
I think that your proposal is basically the same as mine.  I had one other gizmo: a variety of USB MIDI hardware just calls itself the generic "USB MIDI Device".  We'd like to be able to rename that.  Since we can rename the device name, I had added:

* If the endpoint name is "USB MIDI Device" and the device name is non-null, use the device name.

**Comment by brunchboy** [comment]
I saw that, but I did not think it made sense to try to interpret the meaning of non-`null` endpoint names. There is no way we could predict all the ways that different manufacturers might give their devices useless endpoint names. That’s why I say we should honor the name that the user assigns the device in the Audio MIDI Setup control panel, and always show it in some way if it is not `null`.

**Comment by eclab** [comment]
That one's not a big deal, I agree.    You could always rename the device and then you'd have "USB MIDI Device (My Device Name)" or whatnot.

I had included it because I noted that several of my hardware devices all called themselves the same "USB MIDI Device".  Either that's a generic name standard, or perhaps Apple's system was naming them this?  At any rate, it didn't seem arbitrary.

**Comment by DerekCook** [comment]
@eclab before I consider all of the above (where in a lot of places we all seem to be in danger of agreeing in most places), can you please:

- Run the latest snapshot above and see what it returns on your system and report back

- Take, for example, your microSampler, edit the Description in the Device panel and run again and see what you get.

- Likewise, can you change the descrption of these "USB MIDI Devices"?

I don't have much time for a detailed response this morning, as I need to high tail it into work early, but, unless I am mistaken, I think there is some confusion on what is going to be returned as the name under the new scheme - but I would like it verified on what is happening on your system.

Just taking one example end point from your log for your microsampler

```
End Point 
  End Point Reference                  3747857
  End Point kMIDIPropertyName          MIDI IN
  End Point kMIDIPropertyModel         microSAMPLER
  End Point kMIDIPropertyManufacturer  KORG INC.
  End Point kMIDIPropertyOffline       0
  End Point kMIDIPropertyUniqueID      1606973723
  End Point kMIDIPropertyDriverVersion 0

  Entity    Reference                  3747856
  Entity    kMIDIPropertyName          Port 1
  Entity    kMIDIPropertyModel         microSAMPLER
  Entity    kMIDIPropertyManufacturer  KORG INC.
  Entity    kMIDIPropertyOffline       0
  Entity    kMIDIPropertyUniqueID      -2016484299
  Entity    kMIDIPropertyDriverVersion 0

  Device    Reference                  3747855
  Device    kMIDIPropertyName          microSAMPLER
  Device    kMIDIPropertyModel         microSAMPLER
  Device    kMIDIPropertyManufacturer  KORG INC.
  Device    kMIDIPropertyOffline       0
  Device    kMIDIPropertyUniqueID      -593357463
  Device    kMIDIPropertyDriverVersion 0

  Number of entities                   2
```
This is end point data for a device with multiple entities - (2)

The name that CoreMIDI4J should return is `microSAMPLER MIDI IN`

Can you please verify that this is what happens when you run the new snapsot?

**Comment by DerekCook** [comment]
Hi

I had some time tonight to compare suggestions. @eclab suggested
```
NAME:
    dname = ""
    ename = ""
    If endpoint.kMIDIPropertyName != null && !device.endpoint.kMIDIPropertyName.equals("USB MIDI Device")
        ename = endpoint.kMIDIPropertyName;
    If device.kMIDIPropertyName != null && !device.kMIDIPropertyName.equals(endpoint.kMIDIPropertyName)
        dname = device.kMIDIPropertyName;
    if (ename.equals("") && dname.equals(""))
        return "USB MIDI Device"
    else if (ename.equals(""))
        return dname;
    else if (dname.equals(""))
        return ename;
    else
        return ename + " (" + dname + ")";
```
It's not dissimilar to what I have done but I think it is over complicated for the examples I have seen on my own system and what the debug info provided. @brunchboy 's suggestion is simpler

```
 If either of the endpoint or device name is `null`, return only the other. (If both are `null`, just return some generic “unknown device” name as already proposed).

* If the endpoint and device name are both present, and both have the same value, return just one of them, to avoid redundancy.

* If the endpoint and device name are both present and different, return something like `Device Name (Endpoint Name)` or `Endpoint Name - Device Name`, whatever seems to make the most sense.

```
I agree with this apart from the final point. I see no reason to return the endpoint name for a device with a single or no entity if you have a device description. If the user has not edited that description then you get the default device name set by the manufacturer (usually the same as the endpoint name), otherwise you get what the user wants.

The time when it makes sense to combine the Device Name and Endpoint name is when the entity count is more than one. In this case append the endpoint name to the device name.

My code is as follows

```
    if ( deviceName != NULL ) {

      CFStringAppend(buildName, (deviceName != NULL ) ? deviceName :  CFSTR("<Unknown Device>") );
    
      if ( numberOfEntities > 1 ) {
      
        CFStringAppend(buildName, CFSTR(" "));
        CFStringAppend(buildName, (entityName != NULL ) ? entityName :  CFSTR("** NULL POINTER for Entiry Name **") );
      
      }
    
    } else {
      
      CFStringAppend(buildName, (endpointName != NULL ) ? endpointName :  CFSTR("<Unknown Device>") );
      
    }
    
```
What this means
  Device Name = NULL AND  End Point Name = NULL THEN name is "Unknown Device"
  Device NAME = NULL AND End Point Name != NULL THEN name is End Point Name
  Device Name != NULL AND Entity count <= 1 THEN name = Device Name
  Device Name != NULL AND Entity count > 1 THEN name = Device Name + Endpoint Name

As an example, under the old code that only used the end point name, my Motif Rack ES would return
`Port 1`
`Port 2`
...
`Port 8`

From those end point descriptions, I don't really know (other than by surmising) that these ports are from my Motif. What if I had another synth using those end point names. How do I tell them apart?

That is why when there are multiple entities, adding the end point name to the device name gives a meaningful description. For my Motif Rack with Device Name `Motif Rack ES` the port names will be
`Motif Rack ES Port 1`
`Motif Rack ES Port 2`
...
`Motif Rack ES Port 8`

For @eclab's microSampler you should get
`microSampler MIDI OUT`
`microSampler SOUND`
`microSampler MIDI IN`
`microSampler KBD/KNOB`

I think these are far more meaningful names than just using the end point name when there are multiple end points, and the implementation is simpler than suggested.

As mentioned, I would be grateful if you could run the snapshot on your system (it still has the native debug info in) and report bck what you get

Latest snapshot reattached here for convenience.

[coremidi4j-1.1-SNAPSHOT.jar.zip](https://github.com/DerekCook/CoreMidi4J/files/1300879/coremidi4j-1.1-SNAPSHOT.jar.zip)

**Comment by brunchboy** [comment]
I agree, Derek, I am eager to see what Sean gets running that snapshot on his system. It may well be that your current approach is perfectly good. The situation I was trying to protect against, which may be a non-issue, is if there are manufacturers of single-entity devices whose device names are utterly generic and indistinguishable, especially if you have multiple such devices connected. In such cases, yes, you can edit the device name if you know how to do that, but you may have a hard time figuring out which is which, or may not know you can do that.

But if we don’t have any such devices today, there is a good argument for waiting until someone reports an issue about one.

**Comment by DerekCook** [comment]
Hi, James

Agreed, but at that point you should be able to edit the descriptions to distinguish them, which is covered by my code in response to #21 . 

I just need to know if the code is working now. Looks good from here.

**Comment by brunchboy** [comment]
Yes, understood! Looks good to me too. And we can add some instructions on the project page telling people how to rename their devices if they need to.

**Comment by eclab** [comment]
I'm posting a new log of attached devices.  There's good news and bad news.  The good news is that all but one of my devices has a rational name now.  The bad news is that the microSAMPLER names aren't right.

I'm adding a new USB device to the log: an Eventide ModFactor pedal.  The names I get displayed are:

microSAMPLER Port 1
microSAMPLER Port 2
USB MIDI Device Port 1
USB MIDI Device Port 2
MIDI Monitor (Untitled)
PreenFM mk2
USB MIDI Device        [the Tascam US 2x2]
ModFactor Pedal

The bad ones here are microSAMPLER Port 1 and microSAMPLER Port 2.  There are four java MIDI Devices presented from the microSAMPLER: two transmitters and two receivers.  The transmitters are called "KBD/KNOB" and "MIDI IN", and the two receivers are called "SOUND" and "MIDI OUT" (I may have mixed some of these up).  But when presented in Java, now they're both called Port 1 and Port 2.

I believe the reason for this is that the current snapshot may be using the *entity* name, not the *endpoint* name.  On the microSAMPLER (and I have no doubt also the MicroKorg etc.) the entity name is not descriptive.  Would it be possible to use the endpoint name instead?

As to only using the device name if there's a single entity, that seems reasonable to me, though what happens if you have one device, one entity, but two endpoints with different names?  For example, imagine the microSAMPLER with only one port, named SOUND and KBD/KNOB at its in and out.  They'd both be just called microSAMPLER.  I guess that's not a big deal, but...

[log.zip](https://github.com/DerekCook/CoreMidi4J/files/1301700/log.zip)

**Comment by DerekCook** [comment]
Thanks. Yes, I've picked up the entity name by mistake whilst doing some clean up. It should indeed be the endpoint name. My bad.

I didn't spot that as my main rack with my Motif Rack ES is in the keyboard room and weighs a ton, so I didn't haul it back in after the change. Should have checked with it as well.

I'll fix that tonight (and check with the Motif) and we should be there.

Regarding your final point: entity end points usually do have the same name, it's the microsampler that seems to do things differently compared to all the other interfaces I have seen. But end points are in/out pairs, so you usually present them in different lists. But I can do a name compare and add the end point name if they are different.

**Comment by eclab** [comment]
> Fair point on your final case. I've not seen that, have you?

No, just thinking out loud.  I've checked the MicroKorg XL manual and it has the same four endpoints as the Microsampler.  But I wonder if the original Microkorg has only two endpoints and one device.  I know someone who owns one: I might be able to check.  It seems an obvious candidate.

**Comment by DerekCook** [comment]
OK. I have changed my text above, so what you quoted has disappeared. But if you can check this original MicroKorg it would be interesting

**Comment by brunchboy** [comment]
It sounds like we’re getting very close! If you want me to test tonight’s version with some other random MIDI hardware I have lying around I can do so as well. This is going to be a good release, my output menu in Beat Link Trigger is already much more meaningful, though I will definitely have to warn my users to go through and reconfigure all their existing triggers, as the names of many outputs will have changed.

**Comment by eclab** [comment]
Looks like I can't get ahold of the MicroKorg for a few days.  I'll get back with you on its results.

**Comment by DerekCook** [comment]
Hi, yes the more devices we can test with between us the better. Mght as well crack any issues now!

I've checked all of my interfaces/devices here, including dragging my Rack with the Motif In to be beside my Mac. Looks like I missed the entity instead of endpoint name issue because the Motif Entity and Endpoint names are the same. Silly mistake, mind! Also caught another small little bug with the extended info I now put into CoreMidiDeviceInfo, so that is fixed as well.

Anyhow, try this one. It still has the debug info uncommented. If it looks good, I will remove the debug info and push the latest code back to Github ready for a release.

I did not have time to do anything other than correct the end point name, as it has been a very long day over here. But I think that will do for now. 

[coremidi4j-1.1-SNAPSHOT.jar.zip](https://github.com/DerekCook/CoreMidi4J/files/1304416/coremidi4j-1.1-SNAPSHOT.jar.zip)

As James says, we will need to make our users aware of name changes (my x.factory librarians warn when they can't find the port name on startup), but I think it is for the better. I certainly find the names more meaningful esp with my Motif and its multiple entities. My Kronos and Montage are the same, but I have not written librarians for them yet. It will certainly avoid future confusion in my setup when they are added!

**Comment by brunchboy** [comment]
The default names of all of my MIDI devices look good. In many cases, far better than they used to. As far as I am concerned, this release is ready to be shared with the world.

**Comment by DerekCook** [comment]
Thanks, James. Good to hear. I'll finish the code off tonight and push to the main branch in preparation.

**Comment by DerekCook** [comment]
Hi. Have commented out the debug code and pushed the changes to the main branch.

**Comment by brunchboy** [comment]
Thanks, Derek. I will wait a bit in the hopes @eclab has a chance to test the latest snapshot, but plan on deploying a release this weekend barring any surprises.

**Comment by eclab** [comment]
On Sep 15, 2017, at 3:00 PM, James Elliott <notifications@github.com> wrote:

> Thanks, Derek. I will wait a bit in the hopes @eclab has a chance to test the latest snapshot, but plan on deploying a release this weekend barring any surprises.

I can test the snapshot on my current machines tonight.  However I have obtained the MicroKorg and although the manual discusses USB, the unit does not have a USB port!  Probably too old a model.  So that's out for testing.

**Comment by eclab** [comment]
Looking good.  My devices display as:

    microSAMPLER KBD/KNOB [in only]  and  microSAMPLER SOUND [out only]
    microSAMPLER MIDI IN [in only] and microSAMPLER MIDI OUT [out only]
    ModFactor Pedal
    USB MIDI Device  [Renamable to Tascam US 2x2]
    Zero MkII Port 1  [both in and out]
    Zero MkII Port 2  [both in and out]
    ModFactor Pedal
    PreenFM mk2

**Comment by DerekCook** [comment]
Great. Thanks for reporting back. Looks like a wrap then, for now, and I need to get back to other things. 

James, when I get five minutes, I will think of some words to explain the name changes that can go into the release notes.

**Comment by brunchboy** [comment]
Thanks again for all this work, Derek! Let me know if you do, otherwise I was going to write something up myself during the release process. We’ll see who gets there first.

**Comment by brunchboy** [comment]
I am working on a fairly detailed explanation of this change, with sections aimed both at developers using our library, and end-users of their applications, since I will then be able to point my Beat Link Trigger and Afterglow users at the right heading, so if you haven’t started yet, @DerekCook, you may want to wait and see if my effort meets your needs as well.

**Comment by DerekCook** [comment]
No, I haven't started yet. Had to be in work today :-( 

So I'll wait and see what you come up with, if you've started.

**Comment by brunchboy** [comment]
That’s fine! It saves us wasting or duplicating work. :) You can take a look at my draft and see what you think, I need to do other things for a bit. https://github.com/DerekCook/CoreMidi4J#device-names

**Comment by brunchboy** [comment]
I’m guessing you didn’t have time to take a look, but I am going to go ahead and do the release. We can change the README whenever and as often as we want if you find things you want added or changed.

**Comment by DerekCook** [comment]
Hi, no I did not get to look at it. It was a long day, and then chilling out and much needed sleep! I've had a quick look at the description this morning, and it is great. Covers most of what I wanted to say, so will just embellish it a little to add in my deltas. Thanks

**Comment by brunchboy** [comment]
Thanks for those improvements!
